Data/Semiring/Tropical.hs

```diff
@@ -0,0 +1,253 @@
+  -- unfortunately this has to take a `Proxy` because
+  -- we don't have visual type applications before GHC 8.0
+#if !MIN_VERSION_base(4,6,0)
+  extremum :: EProxy e -> Extrema
+#else
+  extremum :: Proxy e -> Extrema
+#endif
+
+instance Extremum 'Minima where
+  extremum _ = Minima
+  {-# INLINE extremum #-} -- just to be safe
+
+instance Extremum 'Maxima where
+  extremum _ = Maxima
+  {-# INLINE extremum #-} -- just to be safe
+
+-- | The tropical semiring. @'Tropical' ''Minima' a@ is equivalent to the semiring
+--     \[
+--     (a \cup \{+\infty\}, \oplus, \otimes)
+--     \]
+--     where
+--     \[
+--     x \oplus y = min\{x,y\},
+--     \]
+--     \[
+--     x \otimes y = x + y.
+--     \]
+--
+--   @'Tropical' ''Maxima' a@ is equivalent to the semiring
+--     \[
+--     (a \cup \{-\infty\}, \oplus, \otimes)
+--     \]
+--     where
+--     \[
+--     x \oplus y = max\{x,y\},
+--     \]
+--     \[
+--     x \otimes y = x + y.
+--     \]
+--
+-- In literature, the 'Semiring' instance of the 'Tropical' semiring lifts
+-- the underlying semiring's additive structure. One might ask why this lifting doesn't
+-- instead witness isn't 'Monoid', since we only lift 'zero' and 'plus' - the reason is
+-- that usually the additive structure of a semiring is monotonic, i.e.
+-- @a '+' ('min' b c) == 'min' (a '+' b) (a '+' c)@, but in general this is not true.
+-- For example, lifting 'Monoid.Product' 'Natural.Natural' or 'Monoid.Product' 'Word' into 'Tropical' is lawful,
+-- but 'Monoid.Product' 'Int' is not, lacking distributivity: @(-1) '*' ('min' 0 1) '/=' 'min' ((-1) '*' 0) ((-1) '*' 1)@.
+-- So, we deviate from literature and instead
+-- witness the lifting of a 'Monoid', so the user must take care to ensure
+-- that their implementation of 'mappend' is monotonic.
+data Tropical :: Extrema -> * -> * where
+  Infinity :: Tropical e a
+  Tropical :: a -> Tropical e a
+#if MIN_VERSION_base(4,7,0)  
+  deriving (Typeable)
+#endif
+
+#if MIN_VERSION_base(4,7,0)
+deriving instance (Typeable e, Data a) => Data (Tropical e a)
+#endif
+deriving instance Eq a => Eq (Tropical e a)
+deriving instance Foldable (Tropical e)
+deriving instance Functor (Tropical e)
+deriving instance Read a => Read (Tropical e a)
+deriving instance Show a => Show (Tropical e a)
+deriving instance Traversable (Tropical e)
+
+instance forall e a. (Ord a, Extremum e) => Ord (Tropical e a) where
+  compare Infinity Infinity         = EQ
+#if !MIN_VERSION_base(4,6,0)  
+  compare Infinity _                = case extremum (EProxy :: EProxy e) of
+    Minima -> LT
+    Maxima -> GT
+  compare _ Infinity                = case extremum (EProxy :: EProxy e) of
+    Minima -> GT
+    Maxima -> LT
+#else
+  compare Infinity _                = case extremum (Proxy :: Proxy e) of
+    Minima -> LT
+    Maxima -> GT
+  compare _ Infinity                = case extremum (Proxy :: Proxy e) of
+    Minima -> GT
+    Maxima -> LT
+#endif  
+  compare (Tropical x) (Tropical y) = compare x y
```

Comment:
Would not two separate instances `Ord a => Ord (Tropical Minima a)` and `Ord a => Ord (Tropical Maxima a)` be more neat?

Reply:
Yes, that would be neater. I agree.